vmm/src/lib.rs

```diff
@@ -2248,10 +2364,53 @@ impl Vmm {
         }
 
         // Now we can restore the rest of the VM.
-        if let Some(ref mut vm) = self.vm {
-            vm.restore()
-        } else {
-            Err(VmError::VmNotCreated)
+        // PANIC: won't panic, we just checked that the VM is there.
+        self.vm.vm_mut().unwrap().restore()
+    }
+
+    /// Checks the migration result.
+    ///
+    /// This should be called when the migration thread indicated a state
+    /// change. The function parses the callback and either shuts down the VMM
+    /// on success or keeps the VM and the VMM running on migration failure.
+    fn check_migration_result(&mut self) {
+        // At this point, the thread must be finished.
+        // TODO print nice error message and return error to outer world.
+        let (vm, migration_res) = self
+            .migration_thread_handle
+            .take()
+            .expect("should have thread")
+            .join()
+            .expect("should have joined");
```

Comment:
Note that even with proper error handling we will loose the `Vm` if the thread panics. I guess we are assuming that can never be the case in which case the final `.expect` is alright.

Reply:
you are right - if we fail here, we are screwed anyway.